File: backend/src/types/candidate.ts
// Shapes shared with the ai-service contract. Kept in one place so
// controllers/services can share consistent typing without re-declaring
// the same interfaces.

export interface Skills {
  primary: string[];
  secondary: string[];
}

export interface ExperienceEntry {
  company: string;
  role: string;
  startDate: string;
  endDate: string;
  durationMonths: number;
  description: string;
}

export interface EducationEntry {
  institution: string;
  degree: string;
  field: string;
  year: string;
}

export interface ProjectEntry {
  name: string;
  description: string;
  techStack: string[];
}

/**
 * Full candidate object as returned by ai-service `POST /ai/parse`.
 * Notably missing `resumeFileUrl`, `uploadedAt`, `fileName` — those are
 * filled in by this backend before persistence.
 */
export interface ParsedCandidate {
  id: string;
  name: string;
  email: string;
  phone: string;
  currentRole: string;
  location: string;
  linkedin: string | null;
  github: string | null;
  portfolio: string | null;
  totalExperienceYears: number;
  availability: string;
  overallRating: number;
  skills: Skills;
  experience: ExperienceEntry[];
  education: EducationEntry[];
  projects: ProjectEntry[];
  certifications: string[];
  languages: string[];
  previousCompanies: string[];
  aiSummary: string;
  careerHighlights: string[];
  strengths: string[];
  weaknesses: string[];
  suitableRoles: string[];
  technologyStack: string[];
  resumeText: string;
}

/** Full candidate record as persisted / returned by this backend. */
export interface Candidate extends ParsedCandidate {
  fileName: string;
  resumeFileUrl: string;
  uploadedAt: Date | string;
  jobId: string | null;
  jobMatchScore: number | null;
  jobMatchBreakdown: JobMatchBreakdown | null;
}

/** Projected subset used in list views and search results. */
export interface CandidateSummary {
  id: string;
  name: string;
  currentRole: string;
  totalExperienceYears: number;
  location: string;
  topSkills: string[];
  aiSummary: string;
  overallRating: number;
  uploadedAt: Date | string;
  availability: string;
  jobId: string | null;
  jobTitle: string | null;
  jobMatchScore: number | null;
}

/**
 * Structured intent parsed from a free-text job description by ai-service
 * `POST /ai/jobs/parse`. Mirrors ai-service's `SearchIntent` model -- same
 * shape the deterministic Ranking Agent scores a candidate against,
 * whether that intent came from a one-off search query or (here) a saved
 * Job's description.
 */
export interface JobIntent {
  designation: string | null;
  requiredSkills: string[];
  minExperience: number | null;
  maxExperience: number | null;
  location: string | null;
  industry: string | null;
  education: string | null;
  availability: string | null;
  keywords: string[];
}

export interface JobMatchBreakdown {
  skillMatch: number;
  experienceMatch: number;
  designationMatch: number;
  industryMatch: number;
  educationMatch: number;
  technologyMatch: number;
  locationMatch: number;
  availabilityMatch: number;
  freshnessScore: number;
}

export interface Job {
  id: string;
  title: string;
  description: string;
  intent: JobIntent;
  createdAt: Date | string;
}

export interface Justification {
  matchingSkills: string[];
  relevantExperience: string;
  strongPoints: string[];
  potentialConcerns: string[];
  recommendation: string;
}

export interface SearchResult {
  candidateId: string;
  rank: number;
  matchScore: number;
  breakdown: JobMatchBreakdown;
  candidate: CandidateSummary;
  justification: Justification | null;
}

export interface SearchResponse {
  query: string;
  totalMatches: number;
  results: SearchResult[];
}

export interface ChatResponse {
  sessionId: string;
  reply: string;
  suggestions: string[];
  candidateIds: string[];
  candidates: CandidateSummary[];
}

File: ai-service/models/candidate.py
"""
The shared Candidate JSON contract. Field names below are snake_case (idiomatic
Python) but every model uses `alias_generator=to_camel` + `populate_by_name=True`
so that (a) JSON in/out is camelCase per the contract, and (b) we can still
construct/access models using either the Python name or the camelCase alias.
"""
from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field
from pydantic.alias_generators import to_camel


class CamelModel(BaseModel):
    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
        extra="ignore",
    )


class Skills(CamelModel):
    primary: list[str] = Field(default_factory=list)
    secondary: list[str] = Field(default_factory=list)


class ExperienceItem(CamelModel):
    company: str = ""
    role: str = ""
    start_date: str = ""
    end_date: str = ""
    duration_months: int = 0
    description: str = ""


class EducationItem(CamelModel):
    institution: str = ""
    degree: str = ""
    field: str = ""
    year: str = ""


class ProjectItem(CamelModel):
    name: str = ""
    description: str = ""
    tech_stack: list[str] = Field(default_factory=list)


class Candidate(CamelModel):
    id: str
    name: str = ""
    email: str = ""
    phone: str = ""
    current_role: str = ""
    location: str = ""
    linkedin: str | None = None
    github: str | None = None
    portfolio: str | None = None
    total_experience_years: float = 0.0
    availability: str = "Not Specified"
    overall_rating: int = 0
    skills: Skills = Field(default_factory=Skills)
    experience: list[ExperienceItem] = Field(default_factory=list)
    education: list[EducationItem] = Field(default_factory=list)
    projects: list[ProjectItem] = Field(default_factory=list)
    certifications: list[str] = Field(default_factory=list)
    languages: list[str] = Field(default_factory=list)
    previous_companies: list[str] = Field(default_factory=list)
    ai_summary: str = ""
    career_highlights: list[str] = Field(default_factory=list)
    strengths: list[str] = Field(default_factory=list)
    weaknesses: list[str] = Field(default_factory=list)
    suitable_roles: list[str] = Field(default_factory=list)
    technology_stack: list[str] = Field(default_factory=list)
    resume_text: str = ""


class CandidateSummary(CamelModel):
    id: str
    name: str = ""
    current_role: str = ""
    total_experience_years: float = 0.0
    location: str = ""
    top_skills: list[str] = Field(default_factory=list)
    ai_summary: str = ""
    overall_rating: int = 0
    uploaded_at: str = ""
    availability: str = "Not Specified"


def candidate_to_summary(candidate: dict) -> dict:
    """Build a CandidateSummary-shaped dict from a full candidate dict
    (both already in camelCase form, e.g. as loaded from candidates_store.json)."""
    skills = candidate.get("skills") or {}
    top_skills = list((skills.get("primary") or []))[:8]
    return {
        "id": candidate.get("id"),
        "name": candidate.get("name", ""),
        "currentRole": candidate.get("currentRole", ""),
        "totalExperienceYears": candidate.get("totalExperienceYears", 0.0),
        "location": candidate.get("location", ""),
        "topSkills": top_skills,
        "aiSummary": candidate.get("aiSummary", ""),
        "overallRating": candidate.get("overallRating", 0),
        "uploadedAt": candidate.get("uploadedAt", ""),
        "availability": candidate.get("availability", "Not Specified"),
    }
